<?xml version="1.0"?>
<!-- 
This launch file initializes the scan_and_approach_node which combines
the functionality of move_to_pose_scan and approach_object nodes.

Command to use with interbotix nav stack:
roslaunch interbotix_xslocobot_nav xslocobot_nav_2.launch robot_model:=locobot_wx250s use_lidar:=true localization:=true
-->
<launch>
  <!-- Global Arguments -->
  <arg name="robot_name" default="locobot"/>
  <arg name="rviz_frame" default="map"/>
  <arg name="rvizconfig" default="$(find yolo_vision)/rviz/xslocobot_description_yolo.rviz"/>
  <arg name="use_rviz" default="true" doc="Set to false to disable RViz visualization"/>
  <arg name="use_cpp_yolo" default="false" doc="Set to true to use C++ implementations instead of Python"/>
  <arg name="use_cpp_distance" default="false" doc="Set to true to use C++ implementations instead of Python"/>
  <arg name="use_cpp_object_mapper" default="false" doc="Set to true to use C++ implementations instead of Python"/>
  <arg name="use_cpp_all" default="false" doc="Set to true to use C++ implementations for all nodes"/>


  <!-- Camera Topics -->
  <arg name="rgb_topic" default="/$(arg robot_name)/camera/color/image_raw"/>
  <arg name="depth_topic" default="/$(arg robot_name)/camera/depth/image_rect_raw"/>

  <!-- YOLO Topics -->
  <arg name="model_path" default="$(find yolo_vision)/models/best.pt"/>
  <arg name="annotated_image_topic" default="camera/yolo/annotated_image"/>
  <arg name="bbox_depth_topic" default="camera/yolo/bbox_depth"/>
  <arg name="visualization_topic" default="camera/yolo/distance_image"/>
  <arg name="camera_info_topic" default="camera/color/camera_info"/>
  <arg name="object_marker_topic" default="object_markers"/>

  <!-- Move and scan parameters -->
  <arg name="poses_config" default="$(find move_to_pose)/config/poses.yaml"/>
  <arg name="pose_command" default="all"/>
  <arg name="move_base_topic" default="/$(arg robot_name)/move_base"/>
  <arg name="costmap_topic" default="/$(arg robot_name)/move_base/global_costmap/costmap"/>

  <!-- Launch YOLO node -->
  <node name="yolo_node" pkg="yolo_vision" 
      type="$(eval 'yolo_node_cpp' if (arg('use_cpp_yolo') or arg('use_cpp_all')) else 'yolo_node.py')"
      output="screen" ns="$(arg robot_name)">
  <param name="model_path" value="$(eval arg('model_path').replace('.pt', '.onnx') if (arg('use_cpp_yolo') or arg('use_cpp_all')) else arg('model_path'))"/>
  <param name="image_topic" value="$(arg rgb_topic)"/>
  <param name="annotated_image_topic" value="$(arg annotated_image_topic)"/>
  <!-- Only needed for C++ version -->
  <param if="$(eval arg('use_cpp_yolo') or arg('use_cpp_all'))" name="names_file" value="$(find yolo_vision)/models/coco.names"/>
  </node>

  <!-- Distance Detection Node -->
  <node name="distance_node" pkg="yolo_vision" 
      type="$(eval 'distance_node' if (arg('use_cpp_distance') or arg('use_cpp_all')) else 'distance_node.py')" 
      output="screen" ns="$(arg robot_name)">
    <!-- Topic configurations -->
    <param name="annotated_image_topic" value="$(arg annotated_image_topic)"/>
    <param name="depth_image_topic" value="$(arg depth_topic)"/>
    <param name="bbox_depth_topic" value="$(arg bbox_depth_topic)"/>
    <param name="visualization_topic" value="$(arg visualization_topic)"/>
    
    <!-- Node configurations -->
    <param name="queue_size" value="10"/>
    <param name="sync_slop" value="0.1"/>
    
    <!-- Color detection parameters -->
    <rosparam param="green_lower">[0, 255, 0]</rosparam>
    <rosparam param="green_upper">[0, 255, 0]</rosparam>
  </node>

  <!-- Object Mapper Node -->
  <node name="object_mapper" pkg="yolo_vision" 
      type="$(eval 'object_mapper_node' if (arg('use_cpp_object_mapper') or arg('use_cpp_all')) else 'object_mapper_node.py')" 
      output="screen" ns="$(arg robot_name)">
    <param name="camera_frame" value="$(arg robot_name)/camera_color_optical_frame"/>
    <param name="map_frame" value="map"/>
    <param name="bbox_depth_topic" value="$(arg bbox_depth_topic)"/>
    <param name="camera_info_topic" value="$(arg camera_info_topic)"/>
    <param name="object_marker_topic" value="$(arg object_marker_topic)"/>
  </node>

  <!-- RViz (conditionally launched) -->
  <group if="$(arg use_rviz)">
      <node name="rviz" pkg="rviz" type="rviz" 
            args="-f $(arg rviz_frame) -d $(arg rvizconfig)"
            ns="$(arg robot_name)">
          <remap from="/clicked_point" to="clicked_point"/>
          <remap from="/initialpose" to="initialpose"/>
          <remap from="/move_base_simple/goal" to="move_base_simple/goal"/>
      </node>
  </group>
</launch>